### DISCUSSION TITLE
Is there a way to use a wasm_bindgen::closure to run async code inside of it?

### DISCUSSION THREAD

**Discussion by jsoneaday** [wasm-bindgen/wasm-bindgen#3889] [Q&A]
I have the code shown below. Is there another way to build an async closure so I can run async code? 

```
pub async fn upload_file(&self, file: &File) {
        let fr = FileReader::new().unwrap();
        fr.read_as_array_buffer(&file.slice().unwrap()).unwrap();
        
        let mut on_load = Closure::wrap(Box::new(move |ev: Event| {
            let file_reader: FileReader = ev.target().unwrap().dyn_into().unwrap();
            let bytes = file_reader.result().unwrap();

            let upload_result = self.client.post(format!("{LITESEED_URL}data"))
                .body(bytes.into_serde::<Vec<u8>>().unwrap())
                .send()
                .await;

            match upload_result {
                Ok(resp) => log!("{:?}", resp),
                Err(e) => log!("{e}")
            }
        }) as Box<dyn FnMut(_)>);
        
        fr.set_onload(Some(on_load.as_ref().unchecked_ref()));
        on_load.forget();
    }
```

**Comment by ghost** [comment]
The problem is that it's not async but needs to be?

You can always use [spawn local](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/fn.spawn_local.html) in WASM to run async code on the browser's async task queue.

**Comment by jsoneaday** [comment]
This gets me a bit closer, but then I get an error that I am capturing self into the closure. So then I cloned self.client outside of the closure but then I get a different error but don't understand why since I am doing a move on outer variables. If client on line 64 has moved into the closure am I not implying that I won't use it again, which I am not.

<img width="1701" alt="Screenshot 2024-03-12 at 8 10 46 PM" src="https://github.com/rustwasm/wasm-bindgen/assets/1474935/337fe0ef-6750-48f1-aebe-a646df73da66">

**Comment by ghost** [comment]
Clone the client again before the second move? 

edit: If it's still not completley clear, the closure is a generator for futures which are sent to the browsers async queue.
